### PR TITLE
Refactor `createTestActions`

### PR DESCRIPTION
**Pull request by martijnbastiaan**
The refactor breaks the fold up in to two stages:

  * Building a test tree with initializers, finalizers, and actions with
    meta information: test path, test dependencies, and test status.

  * Collecting initializers and finalizers in a single pass, and
    applying them to every test action.

This removes an accidental quadratic complexity in favor of a linear
one caused by continuously mapping over tests while building that
list of tests. This removes the need for one layer in `Tr`'s monad
stack: `WriterT`. At the same time, it uses dedicated data structures
instead of tuples for ease of reading.

This commit aims to make future extensions easier.

----------------

Note that we need:

[Expose children in foldGroup](https://github.com/UnkindPartition/tasty/commit/7126701c61d6b2d8f5601e0873df92895589e4d2)

In order to build `TGroup`. The refactor alone doesn't warrant this API change of course, it is mainly used in https://github.com/UnkindPartition/tasty/pull/343.

### CODE REVIEW

**Comment by Bodigrim** [comment]
@andreasabel could you possibly review this please? I'm overcommitted at the moment. The breaking change to `foldGroup` is indeed justified by further development in #343.

**Comment by martijnbastiaan** [comment]
I could rework #343 to not rely on this refactor I suppose, but the stack of `Traversal/Writer/Reader/IO ()` is making my head spin a little bit so it would take some time. Of course I'd prefer it if either you or @andreasabel would find the time to review it. But, no real hurry.

File: core/Test/Tasty/Run.hs [hunk 1]
```diff
@@ -257,89 +252,131 @@ instance Show DependencyException where
...
   :: OptionSet
   -> TestTree
-  -> IO ([(Action, TVar Status)], Seq.Seq Finalizer)
+  -> IO ([TestAction Action], Seq Finalizer)
 createTestActions opts0 tree = do
+  -- Folding the test tree reduces it to a 'TestActionTree', which is a simplified
+  -- version of 'TestTree' that only includes the tests to be run, resources needed
+  -- to run them, and meta information needed to watch test progress and calculate
+  -- dependencies in 'resolveDeps'.
+  unresolvedTestTree <-
```
Comment (Bodigrim) [inline core/Test/Tasty/Run.hs:308]: Could you please add a type signature?

File: core/Test/Tasty/Run.hs [hunk 2]
```diff
@@ -257,89 +252,131 @@ instance Show DependencyException where
...
+        testPath = parentPath |> name
+        testAction = executeTest (run opts test) testStatus (lookupOption opts)
+      pure $ TAction (TestAction {..})
+
+    foldResource :: OptionSet -> ResourceSpec a -> (IO a -> Tr) -> Tr
+    foldResource _opts (ResourceSpec doInit doRelease) a = do
+      initVar <- liftIO $ newTVarIO NotCreated
+      testTree <- a (getResource initVar)
+      let ntests = length (collectTests testTree) -- TODO: Store size of collection
+      finishVar <- liftIO $ newTVarIO ntests      --       in data type?
```
Comment (Bodigrim) [inline core/Test/Tasty/Run.hs:336]: Any resolution on TODO?

File: core/Test/Tasty/Run.hs [hunk 3]
```diff
@@ -257,89 +252,131 @@ instance Show DependencyException where
...
-    mb_tests :: Either [[Path]] [(Action, TVar Status)]
-    mb_tests = resolveDeps $ map
-      (\(act, testInfo) ->
-        (act (Seq.empty, Seq.empty), testInfo))
-      tests
-  case mb_tests of
-    Right tests' -> return (tests', fins)
-    Left cycles -> throwIO (DependencyLoop cycles)
+    finalizers = collectFinalizers unresolvedTestTree
+    tests = collectTests (resolveTestActions unresolvedTestTree)
```
Comment (Bodigrim) [inline core/Test/Tasty/Run.hs:314]: Please add type signatures for `finalizers` and `tests`.

**Review by Bodigrim — COMMENTED** [review]
Overall LGTM, just minor remarks. Nicely done, I've been always too scary to look into `Test.Tasty.Run`.

File: core/Test/Tasty/Run.hs (hunk 2 above)
Comment (martijnbastiaan) [reply core/Test/Tasty/Run.hs]: Picked up in `Cache tree size in 'TestActionTree'`. It is slightly unsafe, see documentation of `TGroup`. I tried moving it to its own module doing the safe constructor thingy, but it dragged in a whole slush of types defined in `Run.hs`. That didn't really help readability IMO, so I left it as it currently is.

**Comment by martijnbastiaan** [comment]
> Nicely done, I've been always too scary to look into `Test.Tasty.Run`.

Thanks! It took me a bit to figure out what made it tick. Hopefully the refactor helps future endeavors related to it.

**Review by Bodigrim — APPROVED** [review]

**Comment by Bodigrim** [comment]
@andreasabel would you like to take a look?

**Comment by Bodigrim** [comment]
@andreasabel this is a tough and time-consuming thing to review, but I'm afraid to burn out the contributor with a long wait. If there are no suggestions from your side by May 1, I'll take a risk to go ahead with a merge.

**Comment by Bodigrim** [comment]
Thanks @martijnbastiaan!

**Pull request merged**